include/ginkgo/core/reorder/rcm.hpp

```diff
@@ -0,0 +1,208 @@
+    {
+        return permutation_;
+    }
+
+    /**
+     * Gets the inverse permutation (permutation matrix, output of the
+     * algorithm) of the linear operator.
+     *
+     * @return the inverse permutation (permutation matrix)
+     */
+    std::shared_ptr<const PermutationMatrix> get_inverse_permutation() const
+    {
+        return inv_permutation_;
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * If this parameter is set then an inverse permutation matrix is also
+         * constructed along with the normal permutation matrix.
+         */
+        bool GKO_FACTORY_PARAMETER(construct_inverse_permutation, false);
+
+        /**
+         * This parameter controls the strategy used to determine a starting
+         * vertex.
+         */
+        starting_strategy GKO_FACTORY_PARAMETER(
+            strategy, starting_strategy::pseudo_peripheral);
+    };
+    GKO_ENABLE_REORDERING_BASE_FACTORY(Rcm, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    /**
+     * Generates the permutation matrix and if required the inverse permutation
+     * matrix.
+     */
+    void generate() const;
+
+    explicit Rcm(std::shared_ptr<const Executor> exec)
+        : EnablePolymorphicObject<Rcm, ReorderingBase>(std::move(exec))
+    {}
+
+    explicit Rcm(const Factory *factory, const ReorderingBaseArgs &args)
+        : EnablePolymorphicObject<Rcm, ReorderingBase>(factory->get_executor()),
+          parameters_{factory->get_parameters()}
+    {
+        const auto exec = this->get_executor();
+        // The adjacency matrix has to be square.
+        GKO_ASSERT_IS_SQUARE_MATRIX(args.system_matrix);
+        // This is needed because it does not make sense to call the copy and
+        // convert if the existing matrix is empty.
+        if (!args.system_matrix->get_size()) {
+            adjacency_matrix_ = SparsityMatrix::create(exec);
+        } else {
+            auto tmp =
+                copy_and_convert_to<SparsityMatrix>(exec, args.system_matrix);
+            // This function provided within the Sparsity matrix format removes
+            // the diagonal elements and outputs an adjacency matrix.
+            adjacency_matrix_ = tmp->to_adjacency_matrix();
+        }
+        auto const dim = adjacency_matrix_->get_size();
+        degrees_ = std::unique_ptr<Array<IndexType>>(
+            new Array<IndexType>(exec, dim[0]));
+        permutation_ = PermutationMatrix::create(exec, dim);
+
+        // To make it explicit.
+        inv_permutation_ = nullptr;
+        if (parameters_.construct_inverse_permutation) {
+            inv_permutation_ = PermutationMatrix::create(exec, dim);
+        }
+
+        this->generate();
+    }
+
+private:
+    std::shared_ptr<SparsityMatrix> adjacency_matrix_;
+    std::shared_ptr<PermutationMatrix> permutation_;
+    std::unique_ptr<Array<IndexType>> degrees_;
```

Comment:
Do you specifically want to store a unique ptr for some reason here ?

Reply:
I thought since this is exclusively owned by the class, i. e. no other owning references should be obtainable, that would make sense. If you'd rather use a shared_ptr, so be it.